x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,499 @@
+import org.elasticsearch.xpack.ql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.ql.util.SpatialCoordinateTypes;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+import static org.apache.lucene.document.ShapeField.QueryRelation.DISJOINT;
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.esql.expression.EsqlTypeResolutions.isSpatial;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.CARTESIAN_SHAPE;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.GEO_SHAPE;
+import static org.elasticsearch.xpack.ql.expression.Expressions.name;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.ql.type.DataTypes.KEYWORD;
+import static org.elasticsearch.xpack.ql.type.DataTypes.NULL;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isString;
+
+public abstract class SpatialRelatesFunction extends BinaryScalarFunction implements EvaluatorMapper {
+    protected DataType spatialDataType;
+    protected final boolean useDocValues;
+
+    protected SpatialRelatesFunction(Source source, Expression left, Expression right, boolean useDocValues) {
+        super(source, left, right);
+        this.useDocValues = useDocValues;
+    }
+
+    public abstract ShapeField.QueryRelation queryRelation();
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.BOOLEAN;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        // We determine the spatial data type first, then check if the other expression is of the same type or a string.
+        TypeResolution leftResolution = isSpatial(left(), sourceText(), FIRST);
+        TypeResolution rightResolution = isSpatial(right(), sourceText(), SECOND);
+        // Both are not spatial, then both could be string literals
+        if ((leftResolution.unresolved() || left().dataType().equals(NULL))
+            && (rightResolution.unresolved() || right().dataType().equals(NULL))) {
+            TypeResolution resolution = bothAreStringTypes(left(), right(), sourceText());
+            if (resolution.unresolved() == false) {
+                spatialDataType = CARTESIAN_SHAPE;
+            }
+            return resolution;
+        }
+        // Both are spatial, but one could be a literal spatial function
+        if (leftResolution.resolved() && rightResolution.resolved()) {
+            if (left().foldable() && right().foldable() == false || DataTypes.isNull(left().dataType())) {
```

Comment:
I found very hard to read these conditions, could you add parenthesis to show the order of execution?